File: tests/wasmbackend/tmemory.nim
# Linear-memory WASM backend test: object field loads/stores at the byte offsets
# `jslayout` computes — the SAME layout core the JS backend uses, emitted here as
# WASM `i32.load`/`i32.store`. The driver (`tmemory.js`) lays a `Point` out in the
# shared memory buffer and passes its byte offset (a pointer) in.

type
  Point = object
    x, y, z: int32

proc dot(p: ptr Point): int32 {.exportc.} =
  # p.x @ offset 0, p.y @ 4, p.z @ 8 — each an `i32.load` at base + jslayout offset.
  result = p.x * p.x + p.y * p.y + p.z * p.z

proc setY(p: ptr Point, v: int32) {.exportc.} =
  # p.y @ offset 4 — an `i32.store`.
  p.y = v

proc manhattan(p: ptr Point): int32 {.exportc.} =
  result = p.x + p.y + p.z

File: tests/wasmbackend/tmemory.js
// Driver for tmemory.nim: lay a Point{x,y,z: int32} out in the shared linear
// memory and pass its byte offset (a pointer) to the exported functions. This is
// the same ArrayBuffer the JS backend uses as `mem` — here it is the module's
// imported WebAssembly.Memory.
"use strict";
const fs = require("fs");
const bytes = fs.readFileSync(process.argv[2]);
const memory = new WebAssembly.Memory({ initial: 1 });
WebAssembly.instantiate(bytes, { env: { memory } }).then(r => {
  const e = r.instance.exports;
  const H = new Int32Array(memory.buffer);
  const p = 64;                       // a byte offset well clear of null
  H[p / 4] = 3; H[p / 4 + 1] = 4; H[p / 4 + 2] = 12;   // x=3, y=4, z=12
  console.log("dot(p) =", e.dot(p));                    // 9 + 16 + 144 = 169
  console.log("manhattan(p) =", e.manhattan(p));        // 3 + 4 + 12 = 19
  e.setY(p, 7);
  console.log("after setY(p,7): y =", H[p / 4 + 1]);    // 7
  console.log("dot(p) =", e.dot(p));                    // 9 + 49 + 144 = 202
}).catch(err => { console.error("WASM ERROR:", err.message); process.exit(1); });
